File: src/main/java/net/ccbluex/liquidbounce/file/configs/AccountsConfig.java
/*
 * LiquidBounce Hacked Client
 * A free open source mixin-based injection hacked client for Minecraft using Minecraft Forge.
 * https://github.com/CCBlueX/LiquidBounce/
 */
package net.ccbluex.liquidbounce.file.configs;

import com.google.gson.*;
import me.liuli.elixir.account.CrackedAccount;
import me.liuli.elixir.account.MinecraftAccount;
import me.liuli.elixir.account.MojangAccount;
import me.liuli.elixir.manage.AccountSerializer;
import net.ccbluex.liquidbounce.file.FileConfig;
import net.ccbluex.liquidbounce.file.FileManager;

import java.io.*;
import java.util.ArrayList;
import java.util.List;

public class AccountsConfig extends FileConfig {
    private final List<MinecraftAccount> accounts = new ArrayList<>();

    /**
     * Constructor of config
     *
     * @param file of config
     */
    public AccountsConfig(final File file) {
        super(file);
    }

    /**
     * Load config from file
     *
     * @throws IOException
     */
    @Override
    protected void loadConfig() throws IOException {
        clearAccounts();

        final JsonElement jsonElement = new JsonParser().parse(new BufferedReader(new FileReader(getFile())));

        if (jsonElement instanceof JsonNull)
            return;

        for (final JsonElement accountElement : jsonElement.getAsJsonArray()) {
            final JsonObject accountObject = accountElement.getAsJsonObject();

            try{
                // Import Elixir account format

                accounts.add(AccountSerializer.INSTANCE.fromJson(accountElement.getAsJsonObject()));
            } catch (JsonSyntaxException | IllegalStateException e) {
                // Import old account format

                JsonElement name = accountObject.get("name");
                JsonElement password = accountObject.get("password");
                JsonElement inGameName = accountObject.get("inGameName");

                if (inGameName.isJsonNull() && password.isJsonNull()) {
                    final MojangAccount mojangAccount = new MojangAccount();

                    mojangAccount.setEmail(name.getAsString());
                    mojangAccount.setName(inGameName.getAsString());
                    mojangAccount.setPassword(password.getAsString());

                    accounts.add(mojangAccount);
                } else {
                    final CrackedAccount crackedAccount = new CrackedAccount();

                    crackedAccount.setName(name.getAsString());

                    accounts.add(crackedAccount);
                }
            }
        }
    }

    /**
     * Save config to file
     *
     * @throws IOException
     */
    @Override
    protected void saveConfig() throws IOException {
        final JsonArray jsonArray = new JsonArray();

        for (final MinecraftAccount minecraftAccount : accounts) {
            jsonArray.add(AccountSerializer.INSTANCE.toJson(minecraftAccount));
        }

        final PrintWriter printWriter = new PrintWriter(new FileWriter(getFile()));
        printWriter.println(FileManager.PRETTY_GSON.toJson(jsonArray));
        printWriter.close();
    }

    /**
     * Add cracked account to config
     *
     * @param name of account
     */
    public void addCrackedAccount(final String name) {
        final CrackedAccount crackedAccount = new CrackedAccount();
        crackedAccount.setName(name);

        if (accountExists(crackedAccount))
            return;

        accounts.add(crackedAccount);
    }

    /**
     * Add account to config
     *
     * @param name     of account
     * @param password of password
     */
    public void addMojangAccount(final String name, final String password) {
        final MojangAccount mojangAccount = new MojangAccount();
        mojangAccount.setName(name);
        mojangAccount.setPassword(password);

        if (accountExists(mojangAccount))
            return;

        accounts.add(mojangAccount);
    }

    /**
     * Add account to config
     */
    public void addAccount(final MinecraftAccount account) {
        accounts.add(account);
    }

    /**
     * Remove account from config
     *
     * @param selectedSlot of the account
     */
    public void removeAccount(final int selectedSlot) {
        accounts.remove(selectedSlot);
    }


    /**
     * Removed an account from the config
     *
     * @param account the account
     */
    public void removeAccount(MinecraftAccount account) {
        accounts.remove(account);
    }

    /**
     * Check if the account is already added
     */
    public boolean accountExists(final MinecraftAccount newAccount) {
        for (final MinecraftAccount minecraftAccount : accounts)
            if (minecraftAccount.getClass().getName().equals(newAccount.getClass().getName()) && minecraftAccount.getName().equals(newAccount.getName()))
                return true;
        return false;
    }

    /**
     * Clear all minecraft accounts from alt array
     */
    public void clearAccounts() {
        accounts.clear();
    }

    /**
     * Get Alts
     *
     * @return list of Alt Accounts
     */
    public List<MinecraftAccount> getAccounts() {
        return accounts;
    }
}

File: src/main/java/net/ccbluex/liquidbounce/file/configs/AccountsConfig.kt
/*
 * LiquidBounce Hacked Client
 * A free open source mixin-based injection hacked client for Minecraft using Minecraft Forge.
 * https://github.com/CCBlueX/LiquidBounce/
 */
package net.ccbluex.liquidbounce.file.configs

import com.google.gson.JsonArray
import com.google.gson.JsonNull
import com.google.gson.JsonParser
import com.google.gson.JsonSyntaxException
import me.liuli.elixir.account.CrackedAccount
import me.liuli.elixir.account.MinecraftAccount
import me.liuli.elixir.account.MojangAccount
import me.liuli.elixir.manage.AccountSerializer.fromJson
import me.liuli.elixir.manage.AccountSerializer.toJson
import net.ccbluex.liquidbounce.file.FileConfig
import net.ccbluex.liquidbounce.file.FileManager.PRETTY_GSON
import java.io.*

class AccountsConfig(file: File) : FileConfig(file) {

    val accounts = mutableListOf<MinecraftAccount>()

    /**
     * Load config from file
     *
     * @throws IOException
     */
    @Throws(IOException::class)
    override fun loadConfig() {
        clearAccounts()
        val jsonElement = JsonParser().parse(file.bufferedReader())
        if (jsonElement is JsonNull) return

        for (accountElement in jsonElement.asJsonArray) {
            val accountObject = accountElement.asJsonObject
            try {
                // Import Elixir account format
                accounts += fromJson(accountElement.asJsonObject)
            } catch (e: JsonSyntaxException) {
                // Import old account format
                val name = accountObject["name"]
                val password = accountObject["password"]
                val inGameName = accountObject["inGameName"]
                if (inGameName.isJsonNull && password.isJsonNull) {
                    val mojangAccount = MojangAccount()
                    mojangAccount.email = name.asString
                    mojangAccount.name = inGameName.asString
                    mojangAccount.password = password.asString
                    accounts += mojangAccount
                } else {
                    val crackedAccount = CrackedAccount()
                    crackedAccount.name = name.asString
                    accounts += crackedAccount
                }
            } catch (e: IllegalStateException) {
                val name = accountObject["name"]
                val password = accountObject["password"]
                val inGameName = accountObject["inGameName"]
                if (inGameName.isJsonNull && password.isJsonNull) {
                    val mojangAccount = MojangAccount()
                    mojangAccount.email = name.asString
                    mojangAccount.name = inGameName.asString
                    mojangAccount.password = password.asString
                    accounts += mojangAccount
                } else {
                    val crackedAccount = CrackedAccount()
                    crackedAccount.name = name.asString
                    accounts += crackedAccount
                }
            }
        }
    }

    /**
     * Save config to file
     *
     * @throws IOException
     */
    @Throws(IOException::class)
    override fun saveConfig() {
        val jsonArray = JsonArray()

        for (minecraftAccount in accounts)
            jsonArray.add(toJson(minecraftAccount))

        file.writeText(PRETTY_GSON.toJson(jsonArray))
    }

    /**
     * Add cracked account to config
     *
     * @param name of account
     */
    fun addCrackedAccount(name: String) {
        val crackedAccount = CrackedAccount()
        crackedAccount.name = name

        if (!accountExists(crackedAccount)) accounts += crackedAccount
    }

    /**
     * Add account to config
     *
     * @param name     of account
     * @param password of password
     */
    fun addMojangAccount(name: String, password: String) {
        val mojangAccount = MojangAccount()
        mojangAccount.name = name
        mojangAccount.password = password

        if (!accountExists(mojangAccount)) accounts += mojangAccount
    }

    /**
     * Add account to config
     */
    fun addAccount(account: MinecraftAccount) = accounts.add(account)

    /**
     * Remove account from config
     *
     * @param selectedSlot of the account
     */
    fun removeAccount(selectedSlot: Int) = accounts.removeAt(selectedSlot)

    /**
     * Removed an account from the config
     *
     * @param account the account
     */
    fun removeAccount(account: MinecraftAccount) = accounts.remove(account)

    /**
     * Check if the account is already added
     */
    fun accountExists(newAccount: MinecraftAccount) =
        accounts.any { it::class == newAccount::class && it.name == newAccount.name }

    /**
     * Clear all minecraft accounts from alt array
     */
    fun clearAccounts() = accounts.clear()
}